### PR TITLE
Replace calls to old repo-updater store and types by the new store

### LINKED ISSUE #14710: repo store: Replace all calls to repo-updater's ListRepos with internal/db/repos#List
Extracted from: https://github.com/sourcegraph/sourcegraph/issues/14092

All calls out to the db package from within `repo-updater` should be instrumented so that we keep the existing metrics currently in the `ObservedStore`.

### PR DESCRIPTION
**Pull request by asdine**
This PR removes the repo-updater `Store` type and replaces all calls to that store by the [new one](https://github.com/sourcegraph/sourcegraph/pull/15343).
Services outside of repo-updater now use the `internal/db` [stores ](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/internal/db/stores.go) instead of relying on repo-updater.
All repo-updater [Repo ](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/cmd/repo-updater/repos/types.go?utm_source=share#L557) and [ExternalService ](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/cmd/repo-updater/repos/types.go#L78) types have been removed, and their logic transfered to `types.Repo` and `types.ExternalService`.

Fixes #14710 #14711 #14712

### CODE REVIEW

File: enterprise/cmd/repo-updater/authz/integration_test.go [hunk 1]
```diff
@@ -126,7 +130,7 @@ func TestIntegration_GitHubPermissions(t *testing.T) {
 	}
 
 	permsStore := edb.NewPermsStore(testDB, clock)
-	syncer := NewPermsSyncer(reposStore, permsStore, clock, nil)
+	syncer := NewPermsSyncer(repoStore, permsStore, clock, nil)
```
Comment (asdine) [inline enterprise/cmd/repo-updater/authz/integration_test.go:133]: This test fails because we are now using the `internal/db/RepoStore` type instead of repo-updater store, and this type filters repos using [authzFilter](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/internal/db/repos.go#L310). @unknwon do you have an idea on how I should fix this?

File: cmd/repo-updater/repos/phabricator.go [hunk 2]
```diff
@@ -136,26 +138,28 @@ func (s *PhabricatorSource) makeRepo(repo *phabricator.Repo) (*Repo, error) {
...
+			URI: name,
+			Sources: map[string]*types.SourceInfo{
+				urn: {
+					ID:       urn,
+					CloneURL: cloneURL,
+					// TODO(tsenart): We need a way for admins to specify which URI to
+					// use as a CloneURL. Do they want to use https + shortname, git + callsign
+					// an external URI that's mirrored or observed, etc.
+					// This must be figured out when starting to integrate the new Syncer with this
+					// source.
```
Comment (unknwon) [inline cmd/repo-updater/repos/phabricator.go:158]: Feels like now is a good chance to create an issue for this. Or is this concern still valid? @tsenart

File: cmd/repo-updater/repos/scheduler_test.go [hunk 3]
```diff
@@ -9,6 +9,7 @@ import (
 
 	"github.com/davecgh/go-spew/spew"
 	"github.com/google/go-cmp/cmp"
+	"github.com/sourcegraph/sourcegraph/cmd/frontend/types"
```
Comment (unknwon) [inline cmd/repo-updater/repos/scheduler_test.go:12]:
```suggestion
	"github.com/google/go-cmp/cmp"

	"github.com/sourcegraph/sourcegraph/cmd/frontend/types"
```

File: cmd/repo-updater/repos/sources.go [hunk 4]
```diff
@@ -8,14 +8,15 @@ import (
 
 	"github.com/hashicorp/go-multierror"
 	"github.com/pkg/errors"
+	"github.com/sourcegraph/sourcegraph/cmd/frontend/types"
```
Comment (unknwon) [inline cmd/repo-updater/repos/sources.go:11]:
```suggestion
	"github.com/pkg/errors"

	"github.com/sourcegraph/sourcegraph/cmd/frontend/types"
```

File: cmd/repo-updater/repos/sources_test.go [hunk 5]
```diff
@@ -19,6 +19,7 @@ import (
 	"github.com/dnaeon/go-vcr/recorder"
 	"github.com/google/go-cmp/cmp"
 	"github.com/inconshreveable/log15"
+	"github.com/sourcegraph/sourcegraph/cmd/frontend/types"
```
Comment (unknwon) [inline cmd/repo-updater/repos/sources_test.go:22]:
```suggestion
	"github.com/inconshreveable/log15"

	"github.com/sourcegraph/sourcegraph/cmd/frontend/types"
```

File: cmd/repo-updater/repos/store.go [hunk 6]
```diff
@@ -4,729 +4,191 @@ import (
...
 	"time"
 
+	"github.com/inconshreveable/log15"
 	"github.com/keegancsmith/sqlf"
 	"github.com/lib/pq"
+	"github.com/opentracing/opentracing-go"
+	otlog "github.com/opentracing/opentracing-go/log"
 	"github.com/pkg/errors"
-
+	"github.com/sourcegraph/sourcegraph/cmd/frontend/types"
```
Comment (unknwon) [inline cmd/repo-updater/repos/store.go:15]:
```suggestion

	"github.com/sourcegraph/sourcegraph/cmd/frontend/types"
```

File: cmd/repo-updater/repos/syncer.go [hunk 7]
```diff
@@ -15,7 +15,9 @@ import (
 	otlog "github.com/opentracing/opentracing-go/log"
 	"github.com/pkg/errors"
 	"github.com/prometheus/client_golang/prometheus"
```
Comment (unknwon) [inline cmd/repo-updater/repos/syncer.go:17]:
```suggestion
	"github.com/pkg/errors"

	"github.com/prometheus/client_golang/prometheus"
```

File: cmd/repo-updater/repos/testing.go [hunk 8]
```diff
@@ -2,20 +2,17 @@ package repos
...
-	"reflect"
-	"sort"
 	"testing"
-	"time"
 
-	"github.com/google/go-cmp/cmp"
 	"github.com/hashicorp/go-multierror"
-	"github.com/sourcegraph/sourcegraph/internal/api"
+	"github.com/pkg/errors"
+	"github.com/sourcegraph/sourcegraph/cmd/frontend/types"
```
Comment (unknwon) [inline cmd/repo-updater/repos/testing.go:9]:
```suggestion
	"github.com/pkg/errors"

	"github.com/sourcegraph/sourcegraph/cmd/frontend/types"
```

File: cmd/repo-updater/repoupdater/server.go [hunk 9]
```diff
@@ -13,8 +13,10 @@ import (
 	"github.com/inconshreveable/log15"
 	otlog "github.com/opentracing/opentracing-go/log"
 	"github.com/pkg/errors"
+	"github.com/sourcegraph/sourcegraph/cmd/frontend/types"
```
Comment (unknwon) [inline cmd/repo-updater/repoupdater/server.go:16]:
```suggestion
	"github.com/pkg/errors"

	"github.com/sourcegraph/sourcegraph/cmd/frontend/types"
```

File: enterprise/cmd/repo-updater/authz/integration_test.go (hunk 1 above)
Comment (unknwon) [reply enterprise/cmd/repo-updater/authz/integration_test.go:133]: Thanks for calling out this, it makes me suspect that some parts in the repo-syncing _should_ have been failing because repo-updater did not have `authzFilter` in place and assumed have unrestricted access to all repos. I'll finish reviewing the rest and will get back to this thread.

**Review by unknwon — APPROVED** [review]
This is epic 🎉

File: cmd/frontend/internal/httpapi/github_webhooks.go [hunk 10]
```diff
@@ -3,30 +3,32 @@ package httpapi
...
 	"strconv"
 	"sync"
 
+	"golang.org/x/sync/errgroup"
+
 	gh "github.com/google/go-github/v28/github"
 	"github.com/inconshreveable/log15"
 
-	"github.com/sourcegraph/sourcegraph/cmd/repo-updater/repos"
+	"github.com/sourcegraph/sourcegraph/cmd/frontend/types"
```
Comment (ryanslade) [inline cmd/frontend/internal/httpapi/github_webhooks.go:16]: We should add a followup task to move this package, it feels wrong to be importing code from `/cmd/frontend` in packages that aren't also part of `cmd/frontend`.

I'm not quite sure what a good place for it is though.. might be worth asking in Slack

File: cmd/repo-updater/internal/store_test.go [hunk 11]
```diff
@@ -1,1152 +0,0 @@
-package internal_test
```
Comment (ryanslade) [inline cmd/repo-updater/internal/store_test.go:1]: Are you comfortable that we haven't lost any important tests by removing this? Did we copy them over to the new store?

File: cmd/repo-updater/repos/integration_test.go [hunk 12]
```diff
@@ -29,40 +18,21 @@ func TestIntegration(t *testing.T) {
...
 	userID := insertTestUser(t, db)
 
 	for _, tc := range []struct {
 		name string
-		test func(*testing.T, repos.Store) func(*testing.T)
+		test func(*testing.T, *repos.Store) func(*testing.T)
 	}{
-		{"DBStore/ListExternalServices", testStoreListExternalServices(userID)},
 		{"DBStore/SyncRateLimiters", testSyncRateLimiters},
-		{"DBStore/ListExternalServices/ByRepo", testStoreListExternalServicesByRepos},
```
Comment (ryanslade) [inline cmd/repo-updater/repos/integration_test.go:54]: Have we duplicated these tests in the new store?

File: cmd/repo-updater/repos/observability.go [hunk 13]
```diff
@@ -2,18 +2,12 @@ package repos
```
Comment (ryanslade) [inline cmd/repo-updater/repos/observability.go:2]: Not to myself for when I get to the end.. Did we keep observability on the new store?

File: internal/db/external_services.go [hunk 14]
```diff
@@ -700,7 +700,7 @@ func (e *ExternalServiceStore) list(ctx context.Context, conds []*sqlf.Query, li
 		SELECT id, kind, display_name, config, created_at, updated_at, deleted_at, last_sync_at, next_sync_at, namespace_user_id, unrestricted
 		FROM external_services
 		WHERE (%s)
-		ORDER BY id DESC
+		ORDER BY id ASC
```
Comment (ryanslade) [inline internal/db/external_services.go:703]: Why did this change?

**Review by ryanslade — APPROVED** [review]
Generally it looks good but I'm worries I may have missed some details due to the size.

I'd feel more confident if we could get this out to the dogfood instance before merging as a sanity check before we merge to main on a Friday :)

**Comment by unknwon** [comment]
@asdine I pushed one commit to fix the failing tests of campaigns repo perms, but there are failings elsewhere.

**Comment by asdine** [comment]
Thanks @unknwon ! These tests started failing when I merge with main this morning, I'll handle the rest of the tests :)

File: internal/db/external_services.go (hunk 14 above)
Comment (asdine) [reply internal/db/external_services.go:703]: Now that this store is used everywhere, I think it should have a predictable behavior. It seemed better to change that and fix any failing test rather than keeping it and sort the results.
The remaining question however is if there is a part of the code that expects results to be sorted in descending order and whose behavior is not enforced in tests. Perhaps @keegancsmith would know?

File: enterprise/cmd/repo-updater/authz/integration_test.go (hunk 1 above)
Comment (unknwon) [reply enterprise/cmd/repo-updater/authz/integration_test.go:133]: I believe this is critical/blocker because all stores initialized from repo-updater should have their "context" wrapped with `backend.WithAuthzBypass(context.Background())`.

File: cmd/repo-updater/internal/store_test.go (hunk 11 above)
Comment (asdine) [reply cmd/repo-updater/internal/store_test.go:1]: Yes they are all still here, they have just been moved

File: internal/db/external_services.go (hunk 14 above)
Comment (keegancsmith) [reply internal/db/external_services.go:703]: I think we did `DESC` so new external services would be synced first. Don't know if that is still relevant in the new architecture, but if so seems like a good property to keep.

File: cmd/repo-updater/repos/integration_test.go (hunk 12 above)
Comment (asdine) [reply cmd/repo-updater/repos/integration_test.go:54]: They haven't been duplicated but the new store already contains tests for each method

File: internal/db/external_services.go (hunk 14 above)
Comment (asdine) [reply internal/db/external_services.go:703]: Noted, I'll make it DESC by default. Thanks!

File: enterprise/cmd/repo-updater/authz/integration_test.go (hunk 1 above)
Comment (unknwon) [reply enterprise/cmd/repo-updater/authz/integration_test.go:134]: I've addressed this in https://github.com/sourcegraph/sourcegraph/pull/15530/commits/9e76d314e7e5c0e705a05ecd73dffacf559c632d. Should be good for now, unless there are places I haven't been able to spot.

File: enterprise/internal/campaigns/reconciler.go [hunk 15]
```diff
@@ -302,7 +302,7 @@ func (e *executor) loadAuthenticator(ctx context.Context) (auth.Authenticator, e
 				return nil, nil
 			}
 
-			return nil, ErrMissingCredentials{repo: e.repo.Name}
+			return nil, ErrMissingCredentials{repo: string(e.repo.Name)}
```
Comment (mrnugget) [inline enterprise/internal/campaigns/reconciler.go:305]: I think it would be better to change the `repo` field to be an `api.RepoName`.

File: enterprise/internal/campaigns/reconciler.go [hunk 16]
```diff
@@ -856,8 +856,8 @@ func reopenAfterDetach(ch *campaigns.Changeset) bool {
 	// TODO: What if somebody closed the changeset on purpose on the codehost?
 }
 
-func loadRepo(ctx context.Context, tx RepoStore, id api.RepoID) (*repos.Repo, error) {
-	rs, err := tx.ListRepos(ctx, repos.StoreListReposArgs{IDs: []api.RepoID{id}})
+func loadRepo(ctx context.Context, tx RepoStore, id api.RepoID) (*types.Repo, error) {
+	rs, err := tx.List(ctx, db.ReposListOptions{IDs: []api.RepoID{id}})
```
Comment (mrnugget) [inline enterprise/internal/campaigns/reconciler.go:860]: Not a blocker, but a note: it would be better to use this https://github.com/sourcegraph/sourcegraph/blob/b0009093e72e5722aa03d94e3bffbc5aaf7a579e/internal/db/repos.go#L92

File: enterprise/internal/campaigns/service.go [hunk 17]
```diff
@@ -451,8 +450,8 @@ func (s *Service) FetchUsernameForBitbucketServerToken(ctx context.Context, exte
 		return "", err
 	}
 
-	rstore := repos.NewDBStore(s.store.DB(), sql.TxOptions{})
-	es, err := rstore.ListExternalServices(ctx, repos.StoreListExternalServicesArgs{IDs: []int64{extSvcID}})
+	rstore := db.NewExternalServicesStoreWithDB(s.store.DB())
+	es, err := rstore.List(ctx, db.ExternalServicesListOptions{IDs: []int64{extSvcID}})
```
Comment (mrnugget) [inline enterprise/internal/campaigns/service.go:454]:
```suggestion
	extSvcStore := db.NewExternalServicesStoreWithDB(s.store.DB())
	es, err := extSvcStore.List(ctx, db.ExternalServicesListOptions{IDs: []int64{extSvcID}})
```

File: enterprise/internal/campaigns/service_apply_campaign.go [hunk 18]
```diff
@@ -151,7 +147,6 @@ type repoExternalID struct {
 type changesetRewirer struct {
 	campaign *campaigns.Campaign
 	tx       *Store
-	rstore   repos.Store
```
Comment (mrnugget) [inline enterprise/internal/campaigns/service_apply_campaign.go:154]: Cool!

File: enterprise/internal/campaigns/service_test.go [hunk 19]
```diff
@@ -16,7 +16,7 @@ import (
 	"github.com/sourcegraph/sourcegraph/internal/actor"
 	"github.com/sourcegraph/sourcegraph/internal/api"
 	"github.com/sourcegraph/sourcegraph/internal/campaigns"
-	"github.com/sourcegraph/sourcegraph/internal/db"
+	edb "github.com/sourcegraph/sourcegraph/internal/db"
```
Comment (mrnugget) [inline enterprise/internal/campaigns/service_test.go:19]: Why this change?

File: enterprise/internal/campaigns/store_changeset_specs_test.go [hunk 20]
```diff
@@ -7,20 +7,24 @@ import (
...
 	cmpgn "github.com/sourcegraph/sourcegraph/internal/campaigns"
+	edb "github.com/sourcegraph/sourcegraph/internal/db"
+	"github.com/sourcegraph/sourcegraph/internal/db/dbutil"
 	"github.com/sourcegraph/sourcegraph/internal/extsvc"
 )
 
-func testStoreChangesetSpecs(t *testing.T, ctx context.Context, s *Store, rs repos.Store, clock clock) {
-	repo := ct.TestRepo(t, rs, extsvc.KindGitHub)
-	deletedRepo := ct.TestRepo(t, rs, extsvc.KindGitHub).With(repos.Opt.RepoDeletedAt(clock.now()))
+func testStoreChangesetSpecs(t *testing.T, ctx context.Context, s *Store, db dbutil.DB, clock clock) {
```
Comment (mrnugget) [inline enterprise/internal/campaigns/store_changeset_specs_test.go:18]: I don't think the 3rd parameter is useful anymore if it's not a repo store. You can get the `DB()` from the `*Store.Handle()`.

File: enterprise/internal/campaigns/store_test.go [hunk 21]
```diff
@@ -22,7 +22,7 @@ type testClock struct {
 func (c *testClock) now() time.Time                { return c.t }
 func (c *testClock) add(d time.Duration) time.Time { c.t = c.t.Add(d); return c.t }
 
-type storeTestFunc func(*testing.T, context.Context, *Store, repos.Store, clock)
+type storeTestFunc func(*testing.T, context.Context, *Store, dbutil.DB, clock)
```
Comment (mrnugget) [inline enterprise/internal/campaigns/store_test.go:25]: Yeah, I don't think this makes sense in this form. The idea behind injecting `repos.Store` was to avoid having to initialise it in every test. Now it needs to be initialised _and_ the `dbutil.DB` that's passed in is the same as in `*Store`.

File: enterprise/internal/campaigns/syncer.go [hunk 22]
```diff
@@ -35,23 +38,27 @@ type SyncRegistry struct {
 }
 
 type RepoStore interface {
-	ListExternalServices(context.Context, repos.StoreListExternalServicesArgs) ([]*repos.ExternalService, error)
-	ListRepos(context.Context, repos.StoreListReposArgs) ([]*repos.Repo, error)
+	List(context.Context, db.ReposListOptions) ([]*types.Repo, error)
```
Comment (mrnugget) [inline enterprise/internal/campaigns/syncer.go:41]: Not a blocker, just a note again: this should be the `Get` method, since only used `List` because we didn't have a `Get` method before.

**Review by mrnugget — APPROVED** [review]
Mostly reviewed the campaigns code: looks good to me! Happy to see this land.

Did you manually test this (i.e. create a campaign and see that nothing blows up)? This close to a release cut I'm scared to run into issues due to our database layer changing significantly.

**Comment by asdine** [comment]
> Did you manually test this (i.e. create a campaign and see that nothing blows up)?

@mrnugget No I did not yet, I will do a complete QA before merging

> This close to a release cut I'm scared to run into issues due to our database layer changing significantly.

Perhaps we should wait for the release? There is no rush in merging this, we'll only need to deal with conflicts, cc @tsenart

File: enterprise/internal/campaigns/reconciler.go [hunk 23]
```diff
@@ -155,14 +154,15 @@ func (e *executor) ExecutePlan(ctx context.Context, plan *plan) (err error) {
 		return nil
 	}
 
-	reposStore := repos.NewDBStore(e.tx.Handle().DB(), sql.TxOptions{})
+	reposStore := db.NewRepoStoreWithDB(e.tx.Handle().DB())
+	esStore := db.NewExternalServicesStoreWithDB(e.tx.Handle().DB())
```
[the remark this answers is not in the document]
Comment (asdine) [reply enterprise/internal/campaigns/reconciler.go:158]: Do you mean using the global `db.ExternalServiceStore.WithDB(e.tx.Handle().DB())` method?

[the remark this answers is not in the document]
Comment (mrnugget) [reply enterprise/internal/campaigns/reconciler.go:158]: No, I meant `.With` on the `basestore`:

https://github.com/sourcegraph/sourcegraph/blob/6b125208c395fbeaa2318ac97ef46d5704c3bd84/internal/db/basestore/store.go#L77-L93

It's part of the "minimal" example of how to use a `basestore`:

https://github.com/sourcegraph/sourcegraph/blob/6b125208c395fbeaa2318ac97ef46d5704c3bd84/internal/db/basestore/store.go#L15-L37

And this is `NewWithHandle`:

https://github.com/sourcegraph/sourcegraph/blob/6b125208c395fbeaa2318ac97ef46d5704c3bd84/internal/db/basestore/store.go#L67-L70

**Comment by tsenart** [comment]
> Perhaps we should wait for the release? There is no rush in merging this, we'll only need to deal with conflicts, cc @tsenart

Yep, that makes sense!

File: enterprise/internal/campaigns/reconciler.go (hunk 23 above)
[the remark this answers is not in the document]
Comment (asdine) [reply enterprise/internal/campaigns/reconciler.go:158]: I'm not sure to understand, are you asking if we could implement `With` as part as the `ExternalServiceStore`? If so, it's already done here https://github.com/sourcegraph/sourcegraph/blob/e49a57f0e29e402a6b121bce0c36423e9e529032/internal/db/external_services.go#L36-L62
However, there is no `NewWithHandle` equivalent for that store. If that's what you are talking about, I'll add it in another PR

[the remark this answers is not in the document]
Comment (mrnugget) [reply enterprise/internal/campaigns/reconciler.go:158]: No, my question was:

> shouldn't it now be possible to implement and use the `.With` method and `NewWithHandle` function?

And it seems like `NewWithHandle` doesn't exist yet, so no to that. But `.With` could be used, if I understand correctly, except that it would require instantiating the new stores and _then_ calling `.With`, which doesn't sound like an improvement over the nested `db.NewExternalServicesStoreWithDB(e.tx.Handle().DB())`.

And with `NewWithHandle` it would be this:

```go
db.NewExternalServicesStoreWithHandle(e.tx.Handle())
```

Not sure if it's worth it. Ideally, I think, we'd have something like this:

```go
db.NewExternalServicesStoreWith(e.tx)
```

**Comment by asdine** [comment]
This PR has been split into multiple small PRs and is no longer necessary
